### PR TITLE
feat: Update command argocd app history to support multiple sources

### PR DESCRIPTION
**Pull request by Mangaal**
### Description

This PR update the ```argocd app history``` command to display the history of an application grouped by their sources. The updated output now includes the source URL, along with the application history details such as ID, date, and revision.

### Changes Made

- Modified the ```argocd app history``` command to group the application history by source.
- Included the source URL in the output for each application group.
- Fixed an issue where the command did not display the revision for sources defined in the spec.sources array.

**Previous command output**
```
argocd app history test-3/text-book
   
ID  DATE                            REVISION
3   2024-03-08 13:04:23 +0530 IST   (2158bc9)
4   2024-03-08 13:07:40 +0530 IST  
5   2024-03-08 13:07:42 +0530 IST  
6   2024-03-08 13:09:56 +0530 IST  
7   2024-03-08 13:16:53 +0530 IST  
8   2024-03-08 13:18:31 +0530 IST  
9   2024-03-08 14:09:41 +0530 IST  
10  2024-03-08 14:09:47 +0530 IST  
11  2024-03-08 14:11:40 +0530 IST  
12  2024-03-08 14:44:14 +0530 IST
```

**Output after changes**

```
dist/argocd app history test-3/text-book

SOURCE  https://github.com/Mangaal/text-store.git
ID      DATE                            REVISION
3       2024-03-08 13:04:23 +0530 IST   (2158bc9)
9       2024-03-08 14:09:41 +0530 IST   (c2ceebf)
10      2024-03-08 14:09:47 +0530 IST   (c2ceebf)
11      2024-03-08 14:11:40 +0530 IST   (c2ceebf)
12      2024-03-08 14:44:14 +0530 IST   (051d4ea)

SOURCE  https://github.com/Mangaal/argocd-yaml.git
ID      DATE                            REVISION
4       2024-03-08 13:07:40 +0530 IST   (8331b14)
5       2024-03-08 13:07:42 +0530 IST   (8331b14)
6       2024-03-08 13:09:56 +0530 IST   (8331b14)
7       2024-03-08 13:16:53 +0530 IST   (8331b14)
8       2024-03-08 13:18:31 +0530 IST   (8331b14)
9       2024-03-08 14:09:41 +0530 IST   (8331b14)
10      2024-03-08 14:09:47 +0530 IST   (8331b14)
11      2024-03-08 14:11:40 +0530 IST   (8331b14)
12      2024-03-08 14:44:14 +0530 IST   (8331b14)
```

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

File: cmd/argocd/commands/app.go [hunk 1]
```diff
@@ -2463,13 +2463,44 @@ func printApplicationHistoryIds(revHistory []argoappv1.RevisionHistory) {
...
+				if len(depInfo.Revisions) == len(depInfo.Sources) && len(depInfo.Revisions[i]) >= 7 {
+					rev = fmt.Sprintf("%s (%s)", rev, depInfo.Revisions[i][0:7])
+				}
+				varHistory[sourceInfo.RepoURL] = append(varHistory[sourceInfo.RepoURL], history{
+					id:       depInfo.ID,
+					date:     depInfo.DeployedAt.String(),
+					revision: rev,
+				})
+			}
+		}
```
Comment (ishitasequeira) [inline cmd/argocd/commands/app.go:2496]: As Sources takes priority over Source, the if clause should be something like below:
```
if depInfo.Sources != nil {
...
} else {
...
}
```

File: cmd/argocd/commands/app_test.go [hunk 2]
```diff
@@ -578,7 +581,74 @@ func TestPrintApplicationHistoryTable(t *testing.T) {
...
 
-	expectation := "ID  DATE                           REVISION\n1   0001-01-01 00:00:00 +0000 UTC  1\n2   0001-01-01 00:00:00 +0000 UTC  2\n3   0001-01-01 00:00:00 +0000 UTC  3\n"
+	expectation := "\nSOURCE  test\nID      DATE                           REVISION\n1       0001-01-01 00:00:00 +0000 UTC  1\n2       0001-01-01 00:00:00 +0000 UTC  2\n3       0001-01-01 00:00:00 +0000 UTC  3\n"
+
+	if output != expectation {
+		t.Fatalf("Incorrect print operation output %q, should be %q", output, expectation)
+	}
+}
+
+func TestPrintApplicationHistoryTableWithMultipleSources(t *testing.T) {
```
Comment (ishitasequeira) [inline cmd/argocd/commands/app_test.go:591]: Let's also add a test where the application has both source and sources, but the history command only displays history of sources.

File: cmd/argocd/commands/app.go [hunk 3]
```diff
@@ -2463,13 +2463,44 @@ func printApplicationHistoryIds(revHistory []argoappv1.RevisionHistory) {
...
+		revision string
+	}
+	varHistory := map[string][]history{}
 	for _, depInfo := range revHistory {
-		rev := depInfo.Source.TargetRevision
-		if len(depInfo.Revision) >= 7 {
-			rev = fmt.Sprintf("%s (%s)", rev, depInfo.Revision[0:7])
+		if depInfo.Source.RepoURL != "" {
+			rev := depInfo.Source.TargetRevision
+			if len(depInfo.Revision) >= 7 {
```
Comment (anandf) [inline cmd/argocd/commands/app.go:2475]: Can we avoid this magic number `7` and create a constant like `MAX_ALLOWED_REVISIONS` ?

Comment (Mangaal) [reply cmd/argocd/commands/app.go]: Hi @anandf , I have removed the no 7 and added a var MAX_ALLOWED_REVISIONS

File: cmd/argocd/commands/app_test.go (hunk 2 above)
Comment (Mangaal) [reply cmd/argocd/commands/app_test.go:591]: Hi @ishitasequeira , I have updated the PR with the changes you have suggested.

File: cmd/argocd/commands/app_test.go [hunk 4]
```diff
@@ -578,7 +581,154 @@ func TestPrintApplicationHistoryTable(t *testing.T) {
...
+	})
+
+	expectation := "\nSOURCE  test\nID      DATE                           REVISION\n0       0001-01-01 00:00:00 +0000 UTC  0\n\nSOURCE  test-1\nID      DATE                           REVISION\n1       0001-01-01 00:00:00 +0000 UTC  1a\n2       0001-01-01 00:00:00 +0000 UTC  2a\n3       0001-01-01 00:00:00 +0000 UTC  3a\n\nSOURCE  test-2\nID      DATE                           REVISION\n1       0001-01-01 00:00:00 +0000 UTC  1b\n2       0001-01-01 00:00:00 +0000 UTC  2b\n3       0001-01-01 00:00:00 +0000 UTC  3b\n"
+
+	if output != expectation {
+		t.Fatalf("Incorrect print operation output %q, should be %q", output, expectation)
+	}
+}
+
+func TestPrintApplicationHistoryTableForWhenBothSourcesAndSourceFiledsExist(t *testing.T) {
```
Comment (ishitasequeira) [inline cmd/argocd/commands/app_test.go:670]: The testcase mentions source and sources field but it seems we covered it in above test case. Maybe we can remove this testcase. Rest of the PR looks good.

Comment (Mangaal) [reply cmd/argocd/commands/app_test.go]: Hi @ishitasequeira , I have removed the extra unit test.

**Review by ishitasequeira — APPROVED** [review]
Thanks @Mangaal, the PR LGTM

**Pull request merged**